lib/MusicBrainz/Server/Data/Work.pm

```diff
@@ -425,6 +426,66 @@ sub _find_recording_artists
     }
 }
 
+=method load_median_length
+
+This method will load the median length for the recordings of the work,
+ignoring any recordings marked as partial or as being a medley.
+
+For context, see the only recording-work relationship type, "performance",
+and its attributes:
+https://musicbrainz.org/relationship/a3005666-a872-32c3-ad06-98af558e99b0
+
+=cut
+
+sub load_median_length
+{
+    my ($self, @works) = @_;
+
+    my @ids = map { $_->id } @works;
+    return () unless @ids;
+
+    my %map = object_to_ids(@works);
+    $self->_find_median_length(\@ids, \%map);
+}
+
+sub _find_median_length
+{
+    my ($self, $ids, $map) = @_;
+    return unless @$ids;
+
+    my $query = "
+        SELECT lrw.entity1 AS work_id, median(r.length) AS length
+        FROM l_recording_work lrw
+        JOIN recording r ON lrw.entity0 = r.id
+        WHERE lrw.entity1 IN (" . placeholders(@$ids) . ")
+        AND NOT EXISTS (
+            SELECT lrw2.entity0, COUNT(*)
+            FROM l_recording_work lrw2
+            WHERE lrw2.entity0 = lrw.entity0
+            GROUP BY lrw2.entity0
+            HAVING COUNT(*) > 1
+        )
+        AND NOT EXISTS (
+            SELECT 1
+            FROM link_attribute la
+            JOIN link_attribute_type lat ON la.attribute_type = lat.id
+            WHERE la.link = lrw.link
+            AND lat.gid IN (
+                'd2b63be6-91ec-426a-987a-30b47f8aae2d', -- partial
+                '37da3398-5d1b-4acb-be25-df95e33e423c' -- medley
+            )
```

Comment:
I'd just check `la.attribute_type IN (579, 750)`, since there's not really any technical reason to prefer the gids internally here. The IDs are just as permanent, and we can save a modest couple milliseconds by not joining `link_attribute_type` on works with a huge number of recordings.

I was testing the query against https://musicbrainz.org/work/0c80db24-389e-3620-8e0b-84dc2b7c009a on pink, and  it takes somewhere between 15 and 20 ms to execute, which is very significant for one tiny number on the page, but that kind of work is also an outlier...for comparison, a work with two recordings took about 0.6 ms.

Reply:
That page takes forever to load already so I don't think this query is a huge added issue tbh... Will get rid of that join though.